blocks/http/test/qa_HttpBlock.cpp

```diff
@@ -0,0 +1,209 @@
+
+#include <fmt/format.h>
+#include <fmt/ranges.h>
+
+#include <gnuradio-4.0/testing/FunctionBlocks.hpp>
+#include <gnuradio-4.0/Block.hpp>
+#include <gnuradio-4.0/Buffer.hpp>
+#include <gnuradio-4.0/Graph.hpp>
+#include <gnuradio-4.0/reflection.hpp>
+#include <gnuradio-4.0/Scheduler.hpp>
+
+#include <gnuradio-4.0/http/HttpBlock.hpp>
+
+template<typename T>
+struct fixed_source : public gr::Block<fixed_source<T>, gr::PortOutNamed<T, "out">> {
+    using super_t = gr::Block<fixed_source<T>, gr::PortOutNamed<T, "out">>;
+    T value = 1;
+
+    [[nodiscard]] constexpr auto
+    processOne() noexcept {
+        return value;
+    }
+
+    void
+    trigger() {
+        gr::Message message;
+        message[gr::message::key::Kind] = "custom_kind";
+        super_t::emitMessage(super_t::msgOut, message);
+    }
+};
+
+const boost::ut::suite HttpBlocktests = [] {
+    using namespace boost::ut;
+    using namespace gr;
+    using namespace std::literals;
+    using namespace std::chrono_literals;
+	using namespace gr::testing;
+    "http GET"_test = [] {
+#ifdef __EMSCRIPTEN__
+        std::thread thread{ [&]() {
+            // see ./pre.js for the emscripten server implementation
+            emscripten_run_script("startServer();");
+        } };
+#else
+        httplib::Server server;
+        server.Get("/echo", [](const httplib::Request, httplib::Response &res) { res.set_content("Hello world!", "text/plain"); });
+
+        auto thread = std::thread{ [&server] { server.listen("localhost", 8080); } };
+        server.wait_until_ready();
+#endif
+
+        gr::Graph graph;
+        auto     &source    = graph.emplaceBlock<fixed_source<uint8_t>>();
+        auto     &httpBlock = graph.emplaceBlock<http::HttpBlock<uint8_t>>("http://localhost:8080", "/echo");
+        auto     &sink      = graph.emplaceBlock<InspectSink<pmtv::map_t>>();
+        graph.connect(source, 0, httpBlock, 0);
+        std::ignore = source.msgOut.connect(httpBlock.msgIn);
+        graph.connect(httpBlock, 0, sink, 0);
+        httpBlock.start();
+
+        // make a request
+        source.trigger();
+        httpBlock.processScheduledMessages();
+        pmtv::map_t response;
+        while (response.empty()) {
+            std::this_thread::sleep_for(1ms);
+            std::ignore = source.work(1UZ);
+            auto work   = httpBlock.work(1UZ);
+            work        = sink.work(1UZ);
+            response    = sink.value;
+        }
+        expect(eq(std::get<std::string>(response.at("raw-data")), "Hello world!"sv));
+
+        // change endpoint to something invalid
+        std::ignore = httpBlock.settings().set({ { "endpoint", "/does-not-exist" } });
+        std::ignore = httpBlock.settings().applyStagedParameters();
+        response.clear();
+        source.trigger();
+        httpBlock.processScheduledMessages();
+        while (response.empty()) {
+            std::this_thread::sleep_for(1ms);
+            std::ignore = source.work(1UZ);
```

Comment:
Is `auto work = httpBlock.work(1UZ);` missing here?

Reply:
Yes, but N/A after the scheduler rewrite.